fixes #1213 - email reports of audit changes

shlomizadok opened the pull request.

shlomizadok committed:
fixes #1213 - Get email reports of audit changes

ShimShtein commented on app/views/audit_mailer/summary.html.erb:
```diff
@@ -0,0 +1,28 @@
+<body style="font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif; color: #3f3f3f; background-color: #f1f1f1; padding: 10px 24px">
```
Consider using one of the css to inline style formatters:
https://www.ruby-toolbox.com/categories/Inline_CSS_for_E-Mail

shlomizadok replied:
@ShimShtein -- maybe open a new issue to remove all (and not only this) E-mails to work with premailer gem (or similar)

ShimShtein replied:
@shlomizadok -- you are right, here is its new ID: http://projects.theforeman.org/issues/8797

ShimShtein commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,22 @@
+  helper :audits, :layout
+
+  def summary(options = {})
+    raise ::Foreman::Exception.new(N_("Must specify a user with email enabled")) unless (user=User.find(options[:user])) && user.mail_enabled?
```
This actually creates log clutter:
http://projects.theforeman.org/issues/8795

ohadlevy commented on app/assets/javascripts/application.js:
```diff
@@ -108,6 +108,8 @@ function onContentLoad(){
     $("#fakepassword").remove();
   });
+
+  $('#audit_summary_query_builder .autocomplete-input').scopedSearch({'delay': 250});
```
imho: this should not be in app.js

ohadlevy commented on app/helpers/users_helper.rb:
```diff
@@ -24,4 +24,21 @@ def user_taxonomies_html_options(user)
   end
+
+  def mail_notification_settings(mail_notification)
+    if mail_notification.queryable
```
nitpick: change to

``` ruby
if mail_notification.queryable?
```

ohadlevy commented on app/helpers/users_helper.rb:
```diff
@@ -24,4 +24,21 @@ def user_taxonomies_html_options(user)
+                                     })
+      (_(mail_notification.description) + link_to_function(settings_icon, "$('##{mail_notification.name}_query_builder').toggle()", {:class => 'mail_notification_query'})).html_safe
+    else
+      _(mail_notification.description)
```
can you extract `_(mail_notification.description)` instead of keeping in both if and else sections?

ohadlevy commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,22 @@
+    raise ::Foreman::Exception.new(N_("Must specify a user with email enabled")) unless (user=User.find(options[:user])) && user.mail_enabled?
+    time = options[:time] ? "time > \"#{options[:time]}\"" : 'time > yesterday'
+    query = "#{options[:query]} and #{time}"
+    limit = options[:limit] || 25
```
does limiting to 25 audits is not too little on a busy system? you can easily miss important values?
also, if we reach the limit, do we tell the user there are more audits (with a link to the audits page + relevant search)?

shlomizadok replied:
What do you think should be the default limit? Also, I have planned (yet didn't execute) to add the option for the user to set the limit (e.g., `options[:limit]`) and anyhow, there is a link to the audits index page to view all the audit changes

ShimShtein replied:
IMHO: All the defaults like line limit (25), time limit (yesterday) should be moved to a single config file.

shlomizadok replied:
refactored to latest 50 audit changes, with display of how many have changed (displaying 50 of 154534 audits...) with a link to the audit index page.

dLobatog commented:
@stbenjam might want to take a look

stbenjam commented:
Sorry for the delay...

A couple of thoughts:
- I like being able to give extra options to a notification :+1: 
- The settings button on the notification feels weird -- but it is in the best place given the current design of the page (which was my doing).  I have the aesthetic sense of a potato... I think someone should take a look at this (in a separate issue)
- Needs a consistent header, it should say <b>FOREMAN</b> AUDIT SUMMARY at the top
- No need for the URL in the subject - https://github.com/theforeman/foreman/pull/1911 introduces tools for identifying the source
- We need to ship a `.text.erb` template as well.  HTML-mail only is bad for the 7 of us mutt users left (but also things like accessibility issues - screen readers, etc.).

shlomizadok commented:
@stbenjam Thank you for your comments. I have applied them:
- The settings button: made it much smaller. I also think that in this current layout, it is the best position for it.
- Added a consistent header (FOREMAN AUDIT SUMMARY)
- Removed URL from subject
- Added text template. can you please test?

shlomizadok commented:
[test]

stbenjam commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,16 @@
+      |  <%= audit.username.gsub(_('User'), '') %>                 |
+      |  ( <%= audit.remote_address %> ")"                         |
+      |  <%= audit_action_name audit%> <%= audited_type audit %>:  |
+      |  <%= audit_title(audit) %>                                 |
```
Couldn't the audit title make the box boundries not line up? It looks nice, but because of that I might avoid it?

shlomizadok commented:
@stbenjam - thanks for reviewing. removed the borders.

stbenjam commented on app/views/audit_mailer/summary.html.erb:
```diff
@@ -0,0 +1,30 @@
+           user_search_param = login ? "user = #{login}" : "username = \"#{audit.username}\""
+        %>
+        <b>
+          <%= link_to(audit.username.gsub(_('User'), ''), audits_path(:host => @url.host, :port => @url.port, :only_path => false, :protocol => @url.scheme, :search => user_search_param)) %>
```
Get a traceback here, it seems possible to have empty usernames

```
1.9.3-p448 :008 > Audit.where("username IS NULL").count
2015-03-03 10:35:50 [D]    (7.0ms)  SELECT COUNT(*) FROM "audits" WHERE (username IS NULL)
 => 60
```

domcleal replied:
Also, what's the `gsub(_("User"), "")` about?

shlomizadok replied:
@domcleal - `Audit.last.username` (would return "Admin User") 
Audit seems to add the word "User" to each username

domcleal replied:
That's because the default admin user has firstname=Admin, lastname=User.  That's its full name.

shlomizadok replied:
I know, I'm just copying the [current UI behavior ](https://github.com/theforeman/foreman/blob/6ace062d9dd9f0a529639d8123bf778072dce18e/app/helpers/audits_helper.rb#L82)

domcleal replied:
Okay, don't copy it and please file a bug against the UI.

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,13 @@
+<%= _("Foreman Audit summary") %>
+=================================
+<% if @audits.size > 0 %>
+  <%= _('Displaying %{num_audits} audits of %{total_audits} audits') % {:num_audits => @limit, :total_audits => @count} %>
```
Use n_() and remove the first "audits" word so there aren't two plural/singular variables.

domcleal commented on app/views/audit_mailer/summary.html.erb:
```diff
@@ -0,0 +1,30 @@
+<h2 style="font-weight: normal; text-transform: uppercase; font-size: 120%;"><%= _("<b>Foreman</b> Audit summary").html_safe %></h2>
+<ul style="list-style-type: none;">
+  <% if @audits.size > 0 %>
+    <%= _('Displaying %{num_audits} audits of %{total_audits} audits') % {:num_audits => @limit, :total_audits => @count} %>
```
ditto, n_() and remove first "audits"

domcleal commented on app/mailers/application_mailer.rb:
```diff
@@ -42,4 +42,10 @@ def group_mail(users, options)
+
+  def set_url
+    unless (@url = URI.parse(Setting[:foreman_url])).present?
+      raise ":foreman_url is not set, please configure in the Foreman Web UI (More -> Settings -> General)"
```
How does this occur?  foreman_url isn't a setting that accepts a blank value AFAIK.

(Administer, not More.)

shlomizadok replied:
It seems that :foreman_url does accept a blank value.
![image](https://cloud.githubusercontent.com/assets/433583/6460081/e75ffdc6-c19c-11e4-8fcc-4ca3643cd0db.png)
So we're raising an error that it should be set.
Maybe it is a bug that you can set a blank value on :foreman_url

domcleal replied:
Yeah, that looks like a bug as foreman_url isn't listed in BLANK_ATTRS, it should be tested for presence.  I'd suggest filing + fixing that rather than adding this method (especially as we set a default value).

shlomizadok replied:
@domcleal - proposed a fix for this issue [9613](http://projects.theforeman.org/issues/9613) also in this PR. Left this `raise` for backwards compatibility

domcleal replied:
As 9613 is a very different sort of bug, you should submit it in its own PR, or at the _very_ least in its own commit.

domcleal commented on app/helpers/users_helper.rb:
```diff
@@ -24,4 +24,18 @@ def user_taxonomies_html_options(user)
+
+  def mail_notification_query_builder_settings(mail_notification)
+    mail_notification.queryable? ? link_to_function(icon_text('cog', "", { :class => "#{mail_notification.name}_query_setting",
+                                                                           :title => _('Add query to this notification')}),
```
I wouldn't advise using "query" in the UI, we generally use "search" or "filter".

stbenjam commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,13 @@
+        ( <%= audit.remote_address %> ")"                         
+        <%= audit_action_name audit%> <%= audited_type audit %>:  
+        <%= audit_title(audit) %>                                 
+  <% end %>
```
Add some white space after each audit, otherwise they get all jumped together

stbenjam commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,13 @@
+  <%= _('Displaying %{num_audits} audits of %{total_audits} audits') % {:num_audits => @limit, :total_audits => @count} %>
+  <% @audits.each do |audit| %>
+        <%= audit.username.gsub(_('User'), '') %>                 
+        ( <%= audit.remote_address %> ")"                         
```
Don't think you need the quotes on `")"`, it renders like this:

```

Foreman Audit summary
=================================
  Displaying 50 audits of 3 audits
        Admin                  
        ( 127.0.0.1 ")"                         
        created Domain:  
        asdfsadf                                 
        Admin                  
        ( 127.0.0.1 ")"                         
        updated Host:  
        errata.bitbin.de                                 
        Admin                  
        ( 127.0.0.1 ")"                         
        updated Host:  
        errata.bitbin.de   
```

stbenjam commented:
Otherwise works well

shlomizadok commented:
Split into two commits - one handles fixing empty foreman_url, another the audit mailer

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,21 @@
+  def summary(options = {})
+    raise ::Foreman::Exception.new(N_("Must specify a user with email enabled")) unless (user=User.find(options[:user])) && user.mail_enabled?
+    time     = options[:time] ? "time > \"#{options[:time]}\"" : 'time > yesterday'
+    @query    = "#{options[:query]} and #{time}"
```
If there's no query, this generates a query of `and ...`

shlomizadok replied:
Which I'm fine with, and is "legal" in scoped_search 
![image](https://cloud.githubusercontent.com/assets/433583/7611727/ef26d9e2-f98e-11e4-988e-b33e0a7e7d62.png)

domcleal replied:
I'm not certain we should rely on that behaviour, try and make it a well-formed search term if possible!

shlomizadok replied:
aye, aye Sir!

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,21 @@
+    mail(:to      => user.mail,
+         :from    => Setting["email_reply_address"],
+         :subject => @subject,
+         :date    => Time.now)
```
I don't think :date matters.

shlomizadok replied:
It doesn't really, just to say when [it was sent](http://api.rubyonrails.org/classes/ActionMailer/Base.html#method-i-mail). Yet, this should be changed to `Time.zone.now` to respect timezone

domcleal replied:
Timezone's irrelevant really, it's a generated header that the MUA will convert on display.

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,21 @@
+    @limit   = options[:limit] || 50
+    @audits  = Audit.search_for(@query).limit(@limit)
+    @count   = Audit.search_for(@query).count
+    @subject = _("Audit summary")
```
Translate after setting the locale

domcleal commented on app/views/audit_mailer/summary.html.erb:
```diff
@@ -0,0 +1,31 @@
+    <% @audits.each do |audit| %>
+      <li style="margin: 5px 15px; border-bottom: 1px solid #ccc; padding-bottom: 10px; width: 70%; font-size: 120%;">
+        <%
+           login = audit.user.login rescue nil
```
What's this rescuing?  Should it be `#try`?

shlomizadok replied:
Sometimes Audit is persisted without a User (e.g., user_id is not saved).
In cases like that, `audit.user.login` or `audit.user.try(:login)` will get a `NoMethodError`
The rescue is for those occasions.

domcleal replied:
Is the NoMethodError because audit.user is nil?

shlomizadok replied:
Indeed

domcleal replied:
Right, so that's what .try solves - it returns nil if the object is nil, and calls that method if it is present.

shlomizadok replied:
yeah, I am familiar with `try` method.
The problem: Audit has it's own unique way to detect the user. audit.user may be a User object, or it may be a String (when no User object). When will try audit.user.try(:login) and the user is a String - we'll still get NoMethodError.
See example at http://fpaste.org/221536/52891214/

domcleal commented on app/views/audit_mailer/summary.html.erb:
```diff
@@ -0,0 +1,31 @@
+<ul style="list-style-type: none;">
+  <% if @audits.size > 0 %>
+    <%= n_('Displaying %{num_audits} of %{total_audits} audits',
+           'Displaying %{num_audits} of %{total_audits} audits', @limit) % {:num_audits => @limit, :total_audits => @count} %>
```
Should the plural be determined on `@count` rather than `@limit`?

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+=================================
+<% if @audits.size > 0 %>
+  <%= n_('Displaying %{num_audits} of %{total_audits} audits',
+         'Displaying %{num_audits} of %{total_audits} audits', @limit) % {:num_audits => @limit, :total_audits => @count} %>
```
ditto re pluralisation

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+  <%= n_('Displaying %{num_audits} of %{total_audits} audits',
+         'Displaying %{num_audits} of %{total_audits} audits', @limit) % {:num_audits => @limit, :total_audits => @count} %>
+  <% @audits.each do |audit| %>
+    <%= audit.username.gsub(_('User'), '') %>
```
ditto re missing username

domcleal commented on app/helpers/users_helper.rb:
```diff
@@ -24,4 +24,18 @@ def user_taxonomies_html_options(user)
+  end
+
+  def mail_notification_query_builder_settings(mail_notification)
+    mail_notification.queryable? ? link_to_function(icon_text('cog', "", { :class => "#{mail_notification.name}_query_setting",
```
The cog seems rather hidden to me, could we perhaps always show the field and associate it with the dropdown menu for the notification, e.g. left indentation or surrounding box

shlomizadok replied:
![image](https://cloud.githubusercontent.com/assets/433583/7617725/b4bda2b4-f9b5-11e4-9aa5-0c4e87ca24cc.png)
It now looks like this, thoughts?

domcleal commented on app/views/users/_audit_summary_query_builder.html.erb:
```diff
@@ -0,0 +1,3 @@
+<div id="audit_summary_query_builder" class="input-group" style="display: none;">
+  <%= text_f(f, :mail_query, {:label => _('Query for %{mailer}') % {:mailer => mailer.humanize }, :id => 'search',:class => "autocomplete-input", :data => {:url => auto_complete_search_audits_path} }) %>
```
Unsure if this is your bug or not, but I see two clear icons:

![screenshot from 2015-05-13 14 35 38](https://cloud.githubusercontent.com/assets/482089/7611605/6062f17a-f97d-11e4-9849-331cf84236f6.png)

shlomizadok replied:
My bug. Fixed.

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,21 @@
+    raise ::Foreman::Exception.new(N_("Must specify a user with email enabled")) unless (user=User.find(options[:user])) && user.mail_enabled?
+    time     = options[:time] ? "time > \"#{options[:time]}\"" : 'time > yesterday'
+    @query    = "#{options[:query]} and #{time}"
+    @limit   = options[:limit] || 50
```
This seems rather arbitrary, we should probably put a warning in the message if the limit has been reached, and the default should probably be `Setting[:entries_per_page]`

domcleal commented on app/views/audit_mailer/summary.html.erb:
```diff
@@ -0,0 +1,31 @@
+      <li style="margin: 5px 15px; border-bottom: 1px solid #ccc; padding-bottom: 10px; width: 70%; font-size: 120%;">
+        <%
+           login = audit.user.login rescue nil
+           user_search_param = login ? "user = #{login}" : "username = \"#{audit.username}\""
```
`login` should be quoted if audit.username has been.  `%Q{username = "#{audit.username}"}` might be clearer in these cases to avoid escaping.

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+<%= _("Foreman Audit summary") %>
```
lower case "Audit" for sentence case

domcleal commented:
Looks in reasonable shape, thanks for rebasing.  It does need some tests to make sure it can render, have a look at the other mailer unit tests.

I'd be happy to take the setting URL validation change in another PR to remove it from here.

shlomizadok commented:
@domcleal - thanks for reviewing and commenting. I have addressed all of your comments and added some unit tests.
UI - Changed to this -> 
![image](https://cloud.githubusercontent.com/assets/433583/7617725/b4bda2b4-f9b5-11e4-9aa5-0c4e87ca24cc.png)
Does this look better than the cog?

dLobatog commented on app/mailers/application_mailer.rb:
```diff
@@ -42,4 +42,10 @@ def group_mail(users, options)
+
+  def set_url
+    unless (@url = URI.parse(Setting[:foreman_url])).present?
+      raise ":foreman_url is not set, please configure in the Foreman Web UI (Administer -> Settings -> General)"
```
Raising a Foreman::Exception instead of StandardError is going to be helpful when debugging issues, also we get ERF numbers etc..

dLobatog commented:
@shlomizadok Tests failing, otherwise :+1:

shlomizadok commented:
Who's tests are failing? :smiling_imp:

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+<%= _("Foreman audit summary") %>
+=================================
+<% if @audits.size > 0 %>
+  <%= n_('Displaying %{num_audits} of %{total_audits} audits',
```
singular would be `%{total_audits} audit` now you've changed what variable n_ is based on (as in, "Displaying 1 of 1 audit")

domcleal commented:
Is there anything here to restrict access to the audit reports to users with the view_audits permission?

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,19 @@
+    set_locale_for(user)
+    @query    = options[:query] ? "#{options[:query]} and #{time}" : "#{time}"
+    @limit   = options[:limit] || SETTINGS[:entries_per_page]
+    @audits  = Audit.search_for(@query).limit(@limit)
```
Actually, here might be a better place to control the permissions by using `Audit.authorized(:view_audit_logs)` so we allow users to sign up for the report, but if they can't view audits, nothing happens.

domcleal commented:
> Does this look better than the cog?

Much nicer, thanks!

domcleal commented on app/views/audit_mailer/summary.html.erb:
```diff
@@ -0,0 +1,31 @@
+<h2 style="font-weight: normal; text-transform: uppercase; font-size: 120%;"><%= _("<b>Foreman</b> audit summary").html_safe %></h2>
+<ul style="list-style-type: none;">
+  <% if @audits.size > 0 %>
+    <%= n_('Displaying %{num_audits} of %{total_audits} audits',
```
ditto singular

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+         'Displaying %{num_audits} of %{total_audits} audits', @count) % {:num_audits => @limit, :total_audits => @count} %>
+  <% @audits.each do |audit| %>
+    <%= audit.username %>
+    ( <%= audit.remote_address %> )
```
These lines (8-10) should probably be on the same line as the username so it reads more correctly, it's a bit disjointed spread over four lines when it's usually shown on one.

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+<% if @audits.size > 0 %>
+  <%= n_('Displaying %{num_audits} of %{total_audits} audits',
+         'Displaying %{num_audits} of %{total_audits} audits', @count) % {:num_audits => @limit, :total_audits => @count} %>
+  <% @audits.each do |audit| %>
```
Indentation for conditionals and for loops should be removed in this mailer, because it's text it all shows up.

Edit: or will `<%-` get rid of it?

domcleal replied:
Just add a blank line between the "Displaying" and the list of audits

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+  <% end %>
+<% else %>
+  <b><%= _('No audit changes for this period') %></b>
+<% end %>
```
Add the URL to the full audit list to the bottom of the text part

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+    <%= audit.username %>
+    ( <%= audit.remote_address %> )
+    <%= audit_action_name audit%> <%= audited_type audit %>:
+    <%= audit_title(audit) %>
```
Add the URL to this single audit to each line

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+    ( <%= audit.remote_address %> )
+    <%= audit_action_name audit%> <%= audited_type audit %>:
+    <%= audit_title(audit) %>
+    -
```
Unsure if this is needed if they're all on one line?  Or a blank line might be clearer, e.g.

<pre>
»Foreman audit summary«
=================================
»Displaying 1 of 2 audits«

Admin User (127.0.0.1) updated Host: cobalt.redhat.com
http://foreman.example.com/audit/123

Admin User (127.0.0.1) updated Host: cobalt.redhat.com
http://foreman.example.com/audit/124
</pre>

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+=================================
+<% if @audits.size > 0 %>
+  <%= n_('Displaying %{num_audits} of %{total_audits} audits',
+         'Displaying %{num_audits} of %{total_audits} audits', @count) % {:num_audits => @limit, :total_audits => @count} %>
```
num_audits seems to be blank for me in both text and HTML mailers, I see `Displaying  of 1 audits` but I can't tell why.  Any idea?

shlomizadok replied:
Do you have `SETTINGS[:entries_per_page]` set ?

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+    -
+  <% end %>
+<% else %>
+  <b><%= _('No audit changes for this period') %></b>
```
Remove `<b>` HTML tags from plain text view.

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,19 @@
+    set_url
+    set_locale_for(user)
+    @query    = options[:query] ? "#{options[:query]} and #{time}" : "#{time}"
+    @limit   = options[:limit] || SETTINGS[:entries_per_page]
```
should be `Setting[:entries_per_page]` as it's a db Setting, not a SETTING(.yaml).

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+  <%= _('No audit changes for this period') %>
+<% end %>
+
+<%= link_to(_('Full audits list'), audits_path(:host => @url.host, :port => @url.port, :only_path => false, :protocol => @url.scheme, :search => @query)) %>
```
This generates HTML into the plain text part, you need to generate a full URL (maybe just the audits_path bit, I haven't tested).

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+    -
+  <% end %>
+<% else %>
+  <%= _('No audit changes for this period') %>
```
Indentation here and above is still present, which looks funny.  Remove it from the left or use `<%-`

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,19 @@
+    set_locale_for(user)
+    @query    = options[:query] ? "#{options[:query]} and #{time}" : "#{time}"
+    @limit   = options[:limit] || SETTINGS[:entries_per_page]
+    @audits  = Audit.authorized(:view_audit_logs).search_for(@query).limit(@limit)
```
I don't think .authorized is working here from the rake task as User.current isn't set to the user we're working on.

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,19 @@
+    set_locale_for(user)
+    @query    = options[:query] ? "#{options[:query]} and #{time}" : "#{time}"
+    @limit   = options[:limit] || Setting[:entries_per_page]
+    @audits  = Audit.joins_authorized_as(user, Audit, :view_audit_logs).search_for(@query).limit(@limit)
```
no need for joins here, just use authorized_as(user, :view_audit_logs, Audit) or set User.current (and properly unset it too!)

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+<%= _("Foreman audit summary") %>
+=================================
+<% if @audits.size > 0 %>
+  <%= n_('Displaying %{num_audits} of %{total_audits} audit',
```
Indentation comment applied to all of these lines as well, not just 12

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+         'Displaying %{num_audits} of %{total_audits} audits', @count) % {:num_audits => @limit, :total_audits => @count} %>
+
+  <% @audits.each do |audit| %>
+    <%= audit.username %> ( <%= audit.remote_address %> ) <%= audit_action_name audit%> <%= audited_type audit %>: <%= audit_title(audit) %> <%= audit_path(:id => audit, :host => @url.host, :port => @url.port, :only_path => false, :protocol => @url.scheme) %>
```
Nitpick, excess space around the remote_address here

domcleal commented on app/views/users/_audit_summary_query_builder.html.erb:
```diff
@@ -0,0 +1,3 @@
+<div id="audit_summary_query_builder" class="input-group">
+  <%= text_f(f, :mail_query, {:label => '', :help_inline => _('Build a query for %{mailer}') % {:mailer => mailer.humanize }, :id => 'search',:class => "autocomplete-input", :data => {:url => auto_complete_search_audits_path} }) %>
```
until mailer names are properly i18ned, use humanize.downcase so it isn't capitalised halfway through the sentence

domcleal commented on test/unit/audit_mailer_test.rb:
```diff
@@ -0,0 +1,49 @@
+      # Add 'create' action to audits
+      @host = FactoryGirl.create(:host)
+    end
+    User.current = users :admin
```
Why use as_admin in a block above (which sets User.current) and still leave User.current set here?  Is this line needed or can we delete it?  It's best to avoid setting User.current unilaterally as it can lead to bugs when it's being set in tests but not in real life.

domcleal commented on test/unit/audit_mailer_test.rb:
```diff
@@ -0,0 +1,49 @@
+  test 'Audit mail should display total count of audits' do
+    @options[:time] = '1973-01-13 00:12'
+    count = Audit.all.count
+    assert_includes(AuditMailer.summary(@options).deliver.body.parts.first.body, "Displaying #{SETTINGS[:entries_per_page]} of #{count} audits")
```
I'd be surprised/worried if this passed with SETTINGS, hardcode the value based on what's in fixtures.

domcleal commented on test/unit/audit_mailer_test.rb:
```diff
@@ -0,0 +1,49 @@
+  test 'Audit html mail should include link to query' do
+    @options[:time] = '1970-01-01'
+    @options[:query] = 'action = create'
+    query_should_be = CGI.escape(%(#{@options[:query]} and time > "#{@options[:time]}"))
```
`require 'cgi'` at the top

shlomizadok replied:
:+1:

domcleal commented on test/unit/audit_mailer_test.rb:
```diff
@@ -0,0 +1,49 @@
+class AuditMailerTest <ActionMailer::TestCase
+  def setup
+    disable_orchestration
+    SETTINGS[:entries_per_page] = 10
```
that's why the test below is passing, there's no need to do this

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,19 @@
+    @query   = options[:query] ? "#{options[:query]} and #{time}" : "#{time}"
+    @limit   = options[:limit] || Setting[:entries_per_page]
+    @audits  = Audit.authorized_as(user, :view_audit_logs, Audit).search_for(@query).limit(@limit)
+    @count   = Audit.search_for(@query).count
```
also use .authorized_as here so the count makes sense

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,19 @@
+
+  def summary(options = {})
+    raise ::Foreman::Exception.new(N_("Must specify a user with email enabled")) unless (user=User.find(options[:user])) && user.mail_enabled?
+    time = options[:time] ? %(time > "#{options[:time]}") : 'time > yesterday'
```
use `>=` instead of `>` due to a scoped_search issue for now.  When you do a greater than search, it actually searches for entries of `time` plus an hour (https://github.com/wvanbergen/scoped_search/blob/v3.2.0/lib/scoped_search/query_builder.rb#L151), so you potentially lose an hour of audits in the report.  `>=` is safer, and there's only an incredibly slim chance of an audit being in two reports, which is better.

The ss behaviour makes sense when testing whole days, but I think it's a bug when you have more precise timestamps, so I'll look at fixing it.

domcleal replied:
Note, leave `time > yesterday` for the reason noted in the second paragraph, it works properly with day comparisons, just change the first timestamp based search.

domcleal replied:
https://github.com/wvanbergen/scoped_search/pull/129 for the issue about search precision.

domcleal commented on app/views/audit_mailer/summary.html.erb:
```diff
@@ -0,0 +1,31 @@
+<ul style="list-style-type: none;">
+  <% if @audits.size > 0 %>
+    <%= n_('Displaying %{num_audits} of %{total_audits} audit',
+           'Displaying %{num_audits} of %{total_audits} audits', @count) % {:num_audits => @limit, :total_audits => @count} %>
```
I don't think `@limit` makes sense here, shouldn't it be the number of values in `@audits`?  e.g. "Displaying 10 of 1 audit" makes no sense.

shlomizadok replied:
Maybe of all Audits? (`Audit.all.count`)
So it will show: "Displaying 10 of 234354 audits" ?

domcleal replied:
The total number of audits in this date period makes most sense to me as the second number, else it might seem strange to receive an e-mail saying it's showing you only a few of (large number) audits.

domcleal replied:
The other reason I say that is the purpose of the line to me is to tell me when I'm looking at a limited number of audits in the report, and not the whole set.  If it says "10 of 234354 audits", I'm not sure if there were ten or eleven that day.

shlomizadok replied:
Okay, changed to:

``` ruby
@count   = Audit.search_for(time).count
@limit   = Setting[:entries_per_page] > @count ? @count : Setting[:entries_per_page]
```

So now it will be "displaying 20 of 101 audits" (when audits on time are bigger than per_page)
Or "displaying 5 of 5 audits" (when audits on time are smaller)

domcleal commented on app/views/audit_mailer/summary.text.erb:
```diff
@@ -0,0 +1,15 @@
+       'Displaying %{num_audits} of %{total_audits} audits', @count) % {:num_audits => @limit, :total_audits => @count} %>
+
+<% @audits.each do |audit| %>
+<%= audit.username %> (<%= audit.remote_address %>) <%= audit_action_name audit%> <%= audited_type audit %>: <%= audit_title(audit) %> <%= audit_path(:id => audit, :host => @url.host, :port => @url.port, :only_path => false, :protocol => @url.scheme) %>
```
Nitpick, put the URL in brackets or a newline perhaps, to separate it from the text of the audit entry

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,19 @@
+    time = options[:time] ? %(time >= "#{options[:time]}") : 'time > yesterday'
+    set_url
+    set_locale_for(user)
+    @count   = Audit.search_for(time).count
```
use `.authorized_as` and search for `@query`?

shlomizadok replied:
:+1: now let's hope :cop: will be happy too :)
( :cop: == robucop )

domcleal commented on app/mailers/audit_mailer.rb:
```diff
@@ -0,0 +1,19 @@
+    time = options[:time] ? %(time >= "#{options[:time]}") : 'time > yesterday'
+    set_url
+    set_locale_for(user)
+    @count   = Audit.authorized_as(user, :view_audit_logs, Audit).search_for(@query).count
```
`@query` is nil here, swap the lines around

domcleal commented:
Merged as bc9558f, thanks @shlomizadok.
